pkg/distributor/distributor.go

```diff
@@ -340,129 +376,188 @@ func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*clie
+			validatedSamples += len(ts.Samples)
 		}
+	}
 
-		keys = append(keys, key)
-		validatedTimeseries = append(validatedTimeseries, validatedSeries)
-		validatedSamples += len(ts.Samples)
+	if len(req.Metadata) > 0 {
+		for _, m := range req.Metadata {
+			key := d.tokenForMetadata(userID, m.MetricName)
+
+			err := d.validateMetadata(m.MetricMetadata, userID)
+
+			if err != nil {
+				// TODO: save the error to a partial error?
+				continue
+			}
+
+			metadataKeys = append(metadataKeys, key)
+			validatedMetadata = append(validatedMetadata, m)
+		}
 	}
+
 	receivedSamples.WithLabelValues(userID).Add(float64(validatedSamples))
+	receivedMetadata.WithLabelValues(userID).Add(float64(len(validatedMetadata)))
+
+	var subRing ring.ReadRing
+	subRing = d.ingestersRing
+
+	// Obtain a subring if required
+	if size := d.limits.SubringSize(userID); size > 0 {
+		h := client.HashAdd32(client.HashNew32(), userID)
+		subRing, err = d.ingestersRing.Subring(h, size)
+		if err != nil {
+			return nil, httpgrpc.Errorf(http.StatusInternalServerError, "unable to create subring: %v", err)
+		}
+	}
 
-	if len(keys) == 0 {
-		// Ensure the request slice is reused if there's no series passing the validation.
-		client.ReuseSlice(req.Timeseries)
+	if len(seriesKeys) == 0 && len(metadataKeys) == 0 {
+		// Ensure the request slice is reused if there's no series or metadata passing the validation.
+		client.ReuseSlice(req.Timeseries, req.Metadata)
 
 		return &client.WriteResponse{}, lastPartialErr
 	}
 
 	now := time.Now()
-	if !d.ingestionRateLimiter.AllowN(now, userID, validatedSamples) {
+	//TODO: might be a problem for customers over the tipping point. Maybe increase limits by around 5 - 10%?
+	totalN := validatedSamples + len(validatedMetadata)
+	if !d.ingestionRateLimiter.AllowN(now, userID, totalN) {
 		// Ensure the request slice is reused if the request is rate limited.
-		client.ReuseSlice(req.Timeseries)
+		client.ReuseSlice(req.Timeseries, req.Metadata)
 
 		// Return a 4xx here to have the client discard the data and not retry. If a client
 		// is sending too much data consistently we will unlikely ever catch up otherwise.
 		validation.DiscardedSamples.WithLabelValues(validation.RateLimited, userID).Add(float64(validatedSamples))
-		return nil, httpgrpc.Errorf(http.StatusTooManyRequests, "ingestion rate limit (%v) exceeded while adding %d samples", d.ingestionRateLimiter.Limit(now, userID), numSamples)
+		validation.DiscardedMetadata.WithLabelValues(validation.RateLimited, userID).Add(float64(len(validatedMetadata)))
+		return nil, httpgrpc.Errorf(http.StatusTooManyRequests, "ingestion rate limit (%v) exceeded while adding %d samples and %d metadata", d.ingestionRateLimiter.Limit(now, userID), numSamples, len(req.Metadata))
 	}
 
-	var subRing ring.ReadRing
-	subRing = d.ingestersRing
+	if len(metadataKeys) > 0 {
+		err = ring.DoBatch(ctx, subRing, metadataKeys, func(ingester ring.IngesterDesc, indexes []int) error {
```

Comment:
On `len(metadataKeys) > 0`, I could probably just make `DoBatch` no-op if there are no key to get rid of the indentation.

The question then becomes should we unify the calls? Technically, we can send metadata + samples together if they're going to the same ingester. At the moment (as you can see in the comment below), I'm prioritizing samples - if metadata fails, I no-op. But again, given the current implementation does not sends them together this is not a problem now.